tests/_client.lua

```diff
@@ -0,0 +1,33 @@
+local spawn = require("awful.spawn")
+
+-- This file provide a simple, yet flexible, test client.
+-- It is used to test the `awful.rules`
+
+return function(class, title)
+    local cmd = table.concat {
+        "echo \"",
+        "local lgi = require 'lgi';",
+        "local Gtk = lgi.require('Gtk');",
+        "local GdkPixbuf = lgi.GdkPixbuf;",
+        "Gtk.init();",
+        "local class = '",
+        class or 'test_app',"';",
+        "local window = Gtk.Window {",
+        "    default_width  = 100,",
+        "    default_height = 100,",
+        "    title          = '",
+            title or 'Awesome test client',
+        "'};",
+        "window:set_wmclass(class, class);",
+        "local app = Gtk.Application {",
+        "    application_id = 'org.awesomewm.tests.",class,
+        "'};",
+        "function app:on_activate()",
+        "    window.application = self;",
+        "    window:show_all();",
+        "end;",
+        "app:run {''}\" | lua"
+    }
+
+    spawn.with_shell(cmd)
```

Comment:
I don't like `spawn.with_shell` due to (in theory) needing escaping on its argument (yeah, this is just test code, but still). I know that this is not an issue here, but could you still change this to `spawn({ "lua", "-e", code })` (and `code` is set up like you do right now).